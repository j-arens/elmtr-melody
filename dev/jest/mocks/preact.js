module.exports = {
    unmountComponent: jest.fn(),
};
