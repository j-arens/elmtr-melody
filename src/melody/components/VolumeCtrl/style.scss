.volumeCtrl {
    position: relative;
    color: $color_white;
    cursor: ns-resize;

    &--isDragging {
        cursor: move;
    }

    &__icon {
        max-width: 21px;
        width: 100%;
        height: 100%;
        fill: currentColor;
        overflow: visible !important;
    }

    &__level,
    &__levelOutline {
        fill: transparent;
        stroke-width: 1px;
        transform-origin: center center;
        transform: scale(1.5) rotate(90deg);
        transition: opacity .15s ease;
    }

    &__level {
        stroke: currentColor;
        stroke-dasharray: 75;
        stroke-linecap: round;
    }

    &__levelOutline {
        stroke: currentColor;
        opacity: 0.5;
    }

}

@media screen and (min-width: 1025px) {
    .volumeCtrl {
        &__level,
        &__levelOutline {
            opacity: 0;
        }

        &:hover,
        &:active {
            .volumeCtrl__level {
                opacity: 1;
            }

            .volumeCtrl__levelOutline {
                opacity: 0.5;
            }
        }
    }
}