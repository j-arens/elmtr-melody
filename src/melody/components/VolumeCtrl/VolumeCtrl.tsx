import { DragProps, MelodyDragEvent } from '@components/DragHelper/type';
import FullVolMarkup from '@components/Icon/svg/fullVol';
import MediumVolMarkup from '@components/Icon/svg/mediumVol';
import MutedMarkup from '@components/Icon/svg/muted';
import { ELEMENTOR_NO_DRAG } from '@constants';
import { Action } from '@redux/type';
import * as classnames from 'classnames';
import { Component, h } from 'preact';
const s = require('./style.scss');

type VolDragDirection =
    | 'up'
    | 'down';

interface Props extends DragProps {
    volume: number;
    toggleVolDragging: () => Action;
    updateVolume: (newVol: number) => Action;
}

interface State {
    originalClientY: number;
    lastClientY: number;
    lastDirection: VolDragDirection;
}

const RANGE_SENSITIVITY = 3000;
const CIRCLE_LENGTH = 75;

export default class VolumeCtrl extends Component<Props, State> {
    state = {
        originalClientY: 0,
        lastClientY: 0,
        lastDirection: null,
    };

    componentDidMount() {
        const { onDragStart, onDrag, onDragEnd, toggleVolDragging } = this.props;

        onDragStart(this.dragStart);
        onDrag(this.setVolume);
        onDragEnd(toggleVolDragging);
    }

    dragStart = ({ clientY }: MelodyDragEvent) => {
        this.props.toggleVolDragging();
        this.setState({ originalClientY: clientY });
    }

    setVolume = ({ clientY }: MelodyDragEvent) => {
        const { lastClientY, lastDirection } = this.state;
        let { originalClientY } = this.state;

        if (clientY === lastClientY) {
            return;
        }

        this.setState({ lastClientY: clientY });
        const direction: VolDragDirection = lastClientY > clientY ? 'up' : 'down';

        if (direction !== null && direction !== lastDirection) {
            originalClientY = clientY;
            this.setState({
                lastDirection: direction,
                originalClientY: clientY,
            });
        }

        const { volume, updateVolume } = this.props;
        const delta = originalClientY - clientY;
        const percentChange = (delta / RANGE_SENSITIVITY);
        let newLevel = percentChange + volume;

        if (newLevel < 0) {
            newLevel = 0;
            this.setState({ originalClientY: clientY });
        }

        if (newLevel > 1) {
            newLevel = 1;
            this.setState({ originalClientY: clientY });
        }

        updateVolume(newLevel);
    }

    getDashOffset(): object {
        const { volume } = this.props;
        const offset = CIRCLE_LENGTH - (volume * CIRCLE_LENGTH);

        return {
            strokeDashoffset: `${offset}`,
        };
    }

    getIcon(): JSX.Element {
        const { volume } = this.props;

        if (volume < 0.1) {
            return MutedMarkup;
        }

        if (volume > 0.1 && volume < 0.8) {
            return MediumVolMarkup;
        }

        if (volume > 0.8) {
            return FullVolMarkup;
        }
    }

    render({ setDragRef, isDragging }) {
        const classes = classnames(s.volumeCtrl, {
            [s['volumeCtrl--isDragging']]: isDragging,
        });
        return (
            <div
                class={classes}
                ref={(ref: HTMLElement) => setDragRef(ref)}
            >
                <svg
                    class={`${s.volumeCtrl__icon} ${ELEMENTOR_NO_DRAG}`}
                    xmlns="http://www.w3.org/2000/svg"
                    viewBox="0 0 24 24"
                >
                    <circle class={s.volumeCtrl__levelOutline} cx="12" cy="12" r="12"/>
                    <circle class={s.volumeCtrl__level} style={this.getDashOffset()} cx="12" cy="12" r="12"/>
                    {this.getIcon()}
                </svg>
            </div>
        );
    }
}
