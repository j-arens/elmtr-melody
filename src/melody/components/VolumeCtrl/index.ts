import DragHelper from '@components/DragHelper/';
import { WithOptionalClassName } from '@melody/components/type';
import { toggleVolDragging, updateVolume } from '@redux/actions';
import { State } from '@redux/type';
import { connect } from 'preact-redux';
import { compose } from 'redux';
import VolumeCtrl, { DispatchProps, StateProps } from './VolumeCtrl';

const mapStateToProps = (state: State) => ({
    volume: state.volume,
});

const mapDispatchToProps = (dispatch) => ({
    toggleVolDragging: () => dispatch(toggleVolDragging()),
    updateVolume: (newLevel: number) => dispatch(updateVolume(newLevel)),
});

// @TODO: update preact for better types, make this work
// export default compose(
//     connect<StateProps, DispatchProps, WithOptionalClassName>(mapStateToProps, mapDispatchToProps),
//     DragHelper,
//     VolumeCtrl,
// );

export default connect<
    StateProps,
    DispatchProps,
    WithOptionalClassName
>(mapStateToProps, mapDispatchToProps)(DragHelper(VolumeCtrl));
