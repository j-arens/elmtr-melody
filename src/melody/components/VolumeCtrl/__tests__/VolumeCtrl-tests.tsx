import { h } from 'preact';
import { render } from 'preact-render-to-string';
import VolumeCtrl from '../VolumeCtrl';

describe('VolumeCtrl', () => {
    let mockprops;

    beforeEach(() => {
        mockprops = {
            toggleVolDragging: jest.fn(),
            updateVolume: jest.fn(),
            isDragging: jest.fn(),
            setDragRef: jest.fn(),
            onDragStart: jest.fn(),
            onDrag: jest.fn(),
            onDragEnd: jest.fn(),
        };
    });

    it('renders', () => {
        const component = render(
            <VolumeCtrl
                volume={1}
                isDragging={false}
                className="silverstone"
                {...mockprops}
            />,
        );
        expect(component).toMatchSnapshot();
    });

    it('50% volume is represented with a half outer circle', () => {
        const component = render(
            <VolumeCtrl
                volume={0.5}
                isDragging={false}
                className="silverstone"
                {...mockprops}
            />,
        );
        expect(component).toMatchSnapshot();
    });

    it('full vol renders full vol icon', () => {
        const component = render(
            <VolumeCtrl
                volume={1}
                isDragging={false}
                className="silverstone"
                {...mockprops}
            />,
        );
        expect(component).toMatchSnapshot();
    });

    it('vol <= 0.1 renders muted icon', () => {
        const component = render(
            <VolumeCtrl
                volume={0.1}
                isDragging={false}
                className="silverstone"
                {...mockprops}
            />,
        );
        expect(component).toMatchSnapshot();
    });

    it('vol < 0.8 && > 0.1 renders medium vol icon', () => {
        const component = render(
            <VolumeCtrl
                volume={0.5}
                isDragging={false}
                className="silverstone"
                {...mockprops}
            />,
        );
        expect(component).toMatchSnapshot();
    });
});
