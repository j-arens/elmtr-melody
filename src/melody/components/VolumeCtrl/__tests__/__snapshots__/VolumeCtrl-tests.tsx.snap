// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`VolumeCtrl 50% volume is represented with a half outer circle 1`] = `
<div class="silverstone volumeCtrl volumeCtrl--isDragging" data-cy="vol-ctrl">
  <svg class="volumeCtrl__icon elementor-tab-title" xmlns="http://www.w3.org/2000/svg"
  viewBox="0 0 24 24">
    <circle class="volumeCtrl__levelOutline" cx="12" cy="12" r="12"></circle>
    <circle class="volumeCtrl__level" style="stroke-dashoffset: 37.5;"
    cx="12" cy="12" r="12"></circle>
    <path d="M18.5 12c0-1.77-1.02-3.29-2.5-4.03v8.05c1.48-.73 2.5-2.25 2.5-4.02zM5 9v6h4l5 5V4L9 9H5z"></path>
  </svg>
</div>
`;

exports[`VolumeCtrl full vol renders full vol icon 1`] = `
<div class="silverstone volumeCtrl volumeCtrl--isDragging" data-cy="vol-ctrl">
  <svg class="volumeCtrl__icon elementor-tab-title" xmlns="http://www.w3.org/2000/svg"
  viewBox="0 0 24 24">
    <circle class="volumeCtrl__levelOutline" cx="12" cy="12" r="12"></circle>
    <circle class="volumeCtrl__level" style="stroke-dashoffset: 0;"
    cx="12" cy="12" r="12"></circle>
    <path d="M3 9v6h4l5 5V4L7 9H3zm13.5 3c0-1.77-1.02-3.29-2.5-4.03v8.05c1.48-.73 2.5-2.25 2.5-4.02zM14 3.23v2.06c2.89.86 5 3.54 5 6.71s-2.11 5.85-5 6.71v2.06c4.01-.91 7-4.49 7-8.77s-2.99-7.86-7-8.77z"></path>
  </svg>
</div>
`;

exports[`VolumeCtrl renders 1`] = `
<div class="silverstone volumeCtrl volumeCtrl--isDragging" data-cy="vol-ctrl">
  <svg class="volumeCtrl__icon elementor-tab-title" xmlns="http://www.w3.org/2000/svg"
  viewBox="0 0 24 24">
    <circle class="volumeCtrl__levelOutline" cx="12" cy="12" r="12"></circle>
    <circle class="volumeCtrl__level" style="stroke-dashoffset: 0;"
    cx="12" cy="12" r="12"></circle>
    <path d="M3 9v6h4l5 5V4L7 9H3zm13.5 3c0-1.77-1.02-3.29-2.5-4.03v8.05c1.48-.73 2.5-2.25 2.5-4.02zM14 3.23v2.06c2.89.86 5 3.54 5 6.71s-2.11 5.85-5 6.71v2.06c4.01-.91 7-4.49 7-8.77s-2.99-7.86-7-8.77z"></path>
  </svg>
</div>
`;

exports[`VolumeCtrl vol < 0.8 && > 0.1 renders medium vol icon 1`] = `
<div class="silverstone volumeCtrl volumeCtrl--isDragging" data-cy="vol-ctrl">
  <svg class="volumeCtrl__icon elementor-tab-title" xmlns="http://www.w3.org/2000/svg"
  viewBox="0 0 24 24">
    <circle class="volumeCtrl__levelOutline" cx="12" cy="12" r="12"></circle>
    <circle class="volumeCtrl__level" style="stroke-dashoffset: 37.5;"
    cx="12" cy="12" r="12"></circle>
    <path d="M18.5 12c0-1.77-1.02-3.29-2.5-4.03v8.05c1.48-.73 2.5-2.25 2.5-4.02zM5 9v6h4l5 5V4L9 9H5z"></path>
  </svg>
</div>
`;

exports[`VolumeCtrl vol <= 0.1 renders muted icon 1`] = `
<div class="silverstone volumeCtrl volumeCtrl--isDragging" data-cy="vol-ctrl">
  <svg class="volumeCtrl__icon elementor-tab-title" xmlns="http://www.w3.org/2000/svg"
  viewBox="0 0 24 24">
    <circle class="volumeCtrl__levelOutline" cx="12" cy="12" r="12"></circle>
    <circle class="volumeCtrl__level" style="stroke-dashoffset: 67.5;"
    cx="12" cy="12" r="12"></circle>
    <path d="M7 9v6h4l5 5V4l-5 5H7z"></path>
  </svg>
</div>
`;
